src/execution_plan/ops/op_value_hash_join.c

```diff
@@ -0,0 +1,289 @@
+        if(SIValue_Order(x, v) < 0) left = pos + 1;
+        else right = pos;
+    }
+    return left;
+}
+
+// Performs binary search, returns the rightmost index of a match.
+static int64_t _binarySearchRightmost(Record *array, int32_t array_len, int join_key_idx, SIValue v) {
+    int32_t pos;
+    int32_t left = 0;
+    int32_t right = array_len;
+    while(left < right) {
+        pos = (right + left) / 2;
+        SIValue x = Record_GetScalar(array[pos], join_key_idx);
+        if(SIValue_Order(v, x) < 0) right = pos;
+        else left = pos + 1;
+    }
+    return right -1;
+}
+
+/* Retrive the next intersecting record
+ * if such exists, otherwise returns NULL. */
+static Record _get_intersecting_record(OpValueHashJoin *op) {
+    // No more intersecting records.
+    if(op->intersect_idx == -1 ||
+       op->number_of_intersections == 0) return NULL;
+
+    Record cr = op->cached_records[op->intersect_idx];
+    
+    // Update intersection trackers.
+    op->intersect_idx++;
+    op->number_of_intersections--;
+
+    return cr;
+}
+
+/* Look up first intersecting cached record CR position.
+ * Returns false if intersecting record is found. */
+static bool _set_intersection_idx(OpValueHashJoin *op, SIValue v) {
+    op->intersect_idx = -1;
+    op->number_of_intersections = 0;
+    uint record_count = array_len(op->cached_records);
+
+    int64_t leftmost_idx = _binarySearchLeftmost(op->cached_records, op->join_value_rec_idx, v);
+    if(leftmost_idx >= record_count) return false;
+
+    // Make sure value was found.
+    Record r = op->cached_records[leftmost_idx];
+    SIValue x = Record_GetScalar(r, op->join_value_rec_idx);
+    if(SIValue_Compare(x, v) != 0) return false; // Value wasn't found.
+
+    /* Value was found
+     * idx points to the first intersecting record.
+     * update number_of_intersections to count how many
+     * records share the same value. */
+    op->intersect_idx = leftmost_idx;
+
+    /* Count how many records share the same node.
+     * reduce search space by truncating left bound */
+    int64_t rightmost_idx = _binarySearchRightmost(op->cached_records + leftmost_idx,
+                                                   record_count - leftmost_idx,
+                                                   op->join_value_rec_idx,
+                                                   v);
+    // Compensate index.
+    rightmost_idx += leftmost_idx;
+    // +1 consider rightmost_idx == leftmost_idx.
+    op->number_of_intersections = rightmost_idx - leftmost_idx + 1;
+    assert(op->number_of_intersections > 0);
+
+    return true;
+}
+
+/* Sorts cached records by joined value. */
+void _sort_cached_records(OpValueHashJoin *op) {
+    uint idx = op->join_value_rec_idx;
+    QSORT(Record, op->cached_records,
+          array_len(op->cached_records), RECORD_SORT_ON_ENTRY);
+}
+
+/* Caches all records coming from left branch. */
+void _cache_records(OpValueHashJoin *op) {
+    assert(op->cached_records == NULL);
```

Comment:
I notice that you've been introducing a lot more asserts lately. Some feel superfluous to me - this function is only called `if(op->cached_records == NULL)`, for example. This is just a note, not a criticism - let me know if you come to any conclusions about this kind of design!

Reply:
during the development I wanted to fail fast, some of the asserts can defiantly be removed